### PR TITLE
Clickhouse setup

### PR DESCRIPTION
**Pull request by cavis**
Closes https://github.com/PRX/castle.prx.org/issues/150

Along with [castlehouse](https://github.com/PRX/castlehouse) and [Infrastructure](https://github.com/PRX/Infrastructure/blob/main/spire/templates/shared-clickhouse.yml), this is a proof of concept for BQ -> Google Cloud Storage -> Clickhouse -> Rails.

This sets up:

- The `clickhouse-activerecord` gem, database.yml, and models.
- Some `ClickhouseUtils` tooling to deal with checking if clickhouse is available/accessible.  I suspect we'll need yet more error handling for the charts, but it's a start.
- Some demo pages at `/podcasts/123/rollups_demo` to pull data

### CODE REVIEW

**Review by farski — CHANGES_REQUESTED** [review]
Would like to see some documentation added to the README about developing locally with Clickhouse.

As a proof of concept this all seems reasonable. I think the biggest thing I would like to see change before we call this v1 is the API. Right now I think you need a lot of preexisting knowledge about what data exists in Clickhouse to be able to query it effectively. Some changes to the API could really improve the developer experience for someone that may have a lot of experience with Feeder, but less familiarity with our raw data.

For example, `rollup_zone()` takes a timezone, which makes sense. But I think that is selecting and/or setting the timezone for the resulting data? Conversely, `rollup_day()` doesn't ever take a day or date as an argument; I think the arguments to `rollup_day`, `rollup_week`, etc are all basically the GROUP BY clause (though I guess the reverse, since unlike standard SQL, this seems to aggregate by default, and each facet increases the number of results returned)? If that is the case, I think I would expect something more similar to ActiveRecord, like

```ruby
Rollups::HourlyDownload.bin(:day).where(podcast_id: [123, 456], hour: 2.weeks.ago..1.week.ago).group(:podcast_id, :feed_slug)
```

I think right now, from a naming standpoint, there's also some contention between whether the data in Clickhouse is the rollup (i.e., `Rollups::HourlyDownload` is saying "we roll up data in a bunch of ways, one of which is downloads per hour), or the result of the query is the rollup (i.e., "query some raw data and rollup a bunch of totals for me"). I know that in some ways it's both, but I think right now it's more confusing than it needs to be.

We should also try to be careful and consistent about how those rollup classes are named. `HourlyDownload` and `DailyUnique` fit well together, and make it clear that one is returning downloads and one is returning uniques, but then something like `DailyAgent` is also returning downloads, but that's not obvious.

If you want daily downloads with UA data, you go through `DailyAgent`. If you want daily downloads without UA data, you go through `HourlyDownload` but _bin_ by day. Or if you're in-the-know you can go through `DailyAgent` even though you don't need the UA data and get back a volume of data that more closely matches what you're actually asking for.

I know this approach matches how we've thought about this data for a long time, but it requires the developer to both know what question to ask, and pretty specifically _how to ask it_. I wonder if we could generalize the API to reduce or eliminate that second part. `Rollups::Downloads.bin(:day).where(podcast_id: 3).group(country_code: "US")`. I think I'd rather let the developer ask anything in a general way, and have the model yell if it's not possible (`.group(country_code: "US", agent_name: "Safari")`), and figure out which table(s) it hits if it is possible mostly transparently. I think we have a small enough number of data sets and data-set-specific attributes that doing that wouldn't be too bad. (Or, the API allows you to ask anything, and behind the scenes it queries/joins any tables necessary to answer it.)

I think if we're going down a path where we want to be able to build, or allow users to build, pretty arbitrary reports on most/all of the data we have access to, the API needs to be pretty flexible.

File: config/database.yml [hunk 1]
```diff
@@ -24,66 +24,34 @@ default: &default
...
+    <<: *default
+  clickhouse:
+    <<: *clickhouse
 
 production:
-  <<: *default
+  primary:
+    <<: *default
+  clickhouse:
+    <<: *clickhouse
```
[the remark this answers is not in the document]
Comment (cavis) [reply config/database.yml:57]: Hrm... I think postgres has to be under `primary` for things to work as is.  Or are you saying rename `*default`?

[the remark this answers is not in the document]
Comment (farski) [reply config/database.yml:57]: Ah ok, I didn't realize there was anything expecting a particular value

**Review by farski — APPROVED** [review]
I'm good with merging this, as long as we are treating it mostly as a POC (not Clickhouse itself, just how we're implementing it), and any work done on top of it is also considered POC. Brandon and I are still in the relatively early stages of figuring out what new Metrics will be, even at a fairly high level, so I think we are going to have to be flexible this data layer.

**Pull request merged**